File: Data Structures & Algorithms/n-th-tribonacci-number/submission-3.js
class Solution {
    /**
     * @param {number} n
     * @return {number}
     */
    tribonacci(n) {
        const dp = [0, 1, 1];

        for (let i = 3; i <= n; i++){
            dp[i] = dp[i-1] + dp[i-2] + dp[i-3];
        }

        return dp[n];

        // if (n == 0) return 0;
        // if (n <= 2) return 1; 

        // return this.tribonacci(n-1) + this.tribonacci(n-2) + this.tribonacci(n-3);
    }
}

File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """
            nums = [1, 2, 2, 3, 3, 3], k = 2
        """

        # Approach 1: list: 2001 elements => not ideal

        # Approach 2: Hash map
        #   { number : frequency }
        #   { 1: 1, 2: 2, 3: 3 }

        frequency = {}
        for num in nums:
            if num in frequency:
                frequency[num] = frequency[num] + 1
            else:
                frequency[num] = 1

        # sort dict

        sorted_freq = sorted(frequency.items(), key=lambda frequency: frequency[1] , reverse=True)
        print(sorted_freq)
        
        result = []
        for i in range(k):
            result.append(sorted_freq[i][0])


        return result
